### PR TITLE
[13.0][FIX] product: merge product variants with same attributes

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Make use of merge_records tool to merge records.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by MiquelRForgeFlow** [comment]
I can retarget to v11 branch if you prefer.

**Comment by kos94ok-3D** [comment]
We have not the constrain for combination_indices in pre-migration so maybe better to calculate combination_indices and then group product by product template and combination_indices?

File: addons/product/migrations/13.0.1.2/pre-migration.py [hunk 1]
```diff
@@ -123,6 +124,36 @@ def insert_missing_product_template_attribute_value(env):
...
+        JOIN product_template pt ON pt.id = pp.product_tmpl_id
+        JOIN templates ON (templates.id = pt.id
+            AND templates.combination_indices = pp.combination_indices)"""
+    )
+    templates = {}
+    for product_id, template_id in env.cr.fetchall():
+        if template_id in templates:
+            templates[template_id] += [product_id]
+        else:
+            templates[template_id] = [product_id]
```
Comment (kos94ok-3D) [inline addons/product/migrations/13.0.1.2/pre-migration.py]: I think  it is looks better.
```suggestion
         templates.setdefault(template_id, []).append(product_id)
```

**Review by kos94ok-3D — APPROVED** [review]

**Comment by kos94ok-3D** [comment]
Hi, @MiquelRForgeFlow Could you rebase it?

**Comment by kos94ok-3D** [comment]
Maybe need to be added in a commit? https://github.com/odoo/odoo/pull/32086
And said that it was added in Odoo v12 6 months after the release of Odoo v13.

**Review by kos94ok-3D — APPROVED** [review]
Tested on a real DB

**Review by pedrobaeza — COMMENTED** [review]
Shouldn't you update combination_indices after the merge?

**Comment by MiquelRForgeFlow** [comment]
> Shouldn't you update combination_indices after the merge?

No, why? The combination_indices are the same, they don't change.

File: addons/product/migrations/13.0.1.2/pre-migration.py [hunk 2]
```diff
@@ -161,6 +161,33 @@ def insert_missing_product_template_attribute_value(env):
...
+        JOIN templates ON (templates.id = pt.id
+            AND templates.combination_indices = pp.combination_indices)"""
+    )
+    templates = {}
+    for product_id, template_id in env.cr.fetchall():
+        templates.setdefault(template_id, []).append(product_id)
+    for template_id in templates:
+        openupgrade_merge_records.merge_records(
+            env, 'product.product', templates[template_id][1:],
+            templates[template_id][0], field_spec=None, method='sql',
```
Comment (MiquelRForgeFlow) [inline addons/product/migrations/13.0.1.2/pre-migration.py:187]: The question now is: should I put a field_spec?

Comment (pedrobaeza) [reply addons/product/migrations/13.0.1.2/pre-migration.py:187]: It depends if the standard fields require it or default merge values are enough

Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:187]: Yeah, my best option is `{'default_code': 'first_not_null', 'active': 'or', 'barcode': 'first_not_null', 'volume': 'max', 'weight': 'max'}`. My concern is for custom modules that add float/char/boolean fields, which would need to add more things in this dict. Maybe we could add in openupgradelib a functionality like if the dict contains `'other_fields': 'preserve'`, then every other field is excluded and preserved.

Comment (MiquelRForgeFlow) [reply addons/product/migrations/13.0.1.2/pre-migration.py:187]: https://github.com/OCA/openupgradelib/pull/270

**Comment by pedrobaeza** [comment]
My take on this is that merging complete products is something that shouldn't be corrected at OpenUpgrade level. If they have done such ugly thing, then it must be fixed before doing the migration.

**Comment by MiquelRForgeFlow** [comment]
Ok (don't delete branch)

**Comment by cvinh** [comment]
FYI this PR fixed our issue where we have a product with same attributes
@MiquelRForgeFlow thanks for not having deleted it

**Pull request closed without merging**